src/linsolve/lsmr.jl

```diff
@@ -0,0 +1,214 @@
+
+            v = fact.V[end]
+            α = fact.αs[end]
+
+            # Construct rotation Qhat_{k,2k+1}.
+            αhat = hypot(αbar, alg.λ)
+            chat = αbar / αhat
+            shat = alg.λ / αhat
+
+            # Use a plane rotation (Q_i) to turn B_i to R_i.
+            ρold = ρ
+            ρ = hypot(αhat, β)
+            c = αhat / ρ
+            s = β / ρ
+            θnew = s * α
+            αbar = c * α
+
+            # Use a plane rotation (Qbar_i) to turn R_i^T to R_i^bar.
+            ρbarold = ρbar
+            ζold = ζ
+            θbar = sbar * ρ
+            ρtemp = cbar * ρ
+            ρbar = hypot(cbar * ρ, θnew)
+            cbar = cbar * ρ / ρbar
+            sbar = θnew / ρbar
+            ζ = cbar * ζbar
+            ζbar = -sbar * ζbar
+
+            # Update h, h_hat, x.
+            hbar = add!!(hbar, h, 1, -θbar * ρ / (ρold * ρbarold))
+            h = add!!(h, v, 1, -θnew / ρ)
+            x = add!!(x, hbar, ζ / (ρ * ρbar), 1)
+
+            ##############################################################################
+            ##
+            ## Estimate of ||r||
+            ##
+            ##############################################################################
+
+            # Apply rotation Qhat_{k,2k+1}.
+            βacute = chat * βdd
+            βcheck = -shat * βdd
+
+            # Apply rotation Q_{k,k+1}.
+            βhat = c * βacute
+            βdd = -s * βacute
+
+            # Apply rotation Qtilde_{k-1}.
+            θtildeold = θtilde
+            ρtildeold = hypot(ρdold, θbar)
+            ctildeold = ρdold / ρtildeold
+            stildeold = θbar / ρtildeold
+            θtilde = stildeold * ρbar
+            ρdold = ctildeold * ρbar
+            βd = -stildeold * βd + ctildeold * βhat
+
+            τtildeold = (ζold - θtildeold * τtildeold) / ρtildeold
+            τd = (ζ - θtilde * τtildeold) / ρdold
+            d += abs2(βcheck)
+            normr = sqrt(d + abs2(βd - τd) + abs2(βdd))
+
+            # Estimate ||A||.
+            normA2 += abs2(β)
+            normA = sqrt(normA2)
+            normA2 += abs2(α)
+
+            # Estimate cond(A).
+            maxrbar = max(maxrbar, ρbarold)
+            if length(fact) > 1
+                minrbar = min(minrbar, ρbarold)
+            end
+            condA = max(maxrbar, ρtemp) / min(minrbar, ρtemp)
+
+            ##############################################################################
+            ##
+            ## Test for convergence
+            ##
+            ##############################################################################
+
+            # Compute norms for convergence testing.
+            normAr = abs(ζbar)
+            normx = norm(x)
+
+            # Now use these norms to estimate certain other quantities,
+            # some of which will be small near a solution.
+            test1 = normr / normb
+            test2 = normAr / (normA * normr)
+            test3 = inv(condA)
+
```

Comment:
I'm a bit confused by the placement of these variables. Is there a reason to define these before the logger, and to then use `test1`, `test2` and `test3` below with some unexplained conditions? I think it is a bit more readable to keep that together:
`test3 + 1 <= 1` vs `inv(condA) + 1 <= 1` seems like it reads better

Reply:
Are these different convergence criteria currently used? If not, I would probably streamline them to be compatible with what we use in the other methods, even if this "reduces" the functionality or options to specify the convergence. I will take a stab at this myself.